conda/misc.py

```diff
@@ -177,6 +160,54 @@ def explicit(
     txn.execute()
 
 
+def _get_package_record_from_specs(specs: list[str]) -> Iterable[PackageCacheRecord]:
+    """Given a list of specs, find the corresponding PackageCacheRecord. If
+    some PackageCacheRecords are missing, raise an error.
+    """
+    # group all specs with their package cache record together, and all specs without 
+    # a package cache record together. Will create a dict of the form:
+    # { True: [(spec, record), ...], False: [(spec, None), ...] } 
+    grouped_specs = groupby_to_dict(
+        lambda tup: tup[1] is not None,
+        tuple(
+            [spec, next(PackageCacheData.query_all(spec), None)] for spec in specs
+        ),
+
+    )
+
+    existing = grouped_specs.get(True) or []
+    missing = grouped_specs.get(False) or []
+
+    # Assert that every spec has a PackageCacheRecord
+    if missing and not existing:
+        raise SpecNotFoundInPackageCache("No package cache records found")
+    elif missing:
+        missing_precs_list = ", ".join([str(tup[0]) for tup in missing])
+        raise SpecNotFoundInPackageCache(
+            f"Missing package cache records for: {missing_precs_list}"
+        )
+    return [tup[1] for tup in existing]
```

Comment:
Ah, I glossed over the 1 index, in that case I don't see any reason to use `groupby_to_dict`:

```suggestion
    existing = []
    missing = []
    for spec in specs:
        pcrec = next(PackageCacheData.query_all(spec), None)
        if pcrec is None:
            missing.append(spec)
        else:
            existing.append(pcrec)

    # Assert that every spec has a PackageCacheRecord
    if missing and not existing:
        raise SpecNotFoundInPackageCache("No package cache records found")
    elif missing:
        raise SpecNotFoundInPackageCache(
            f"Missing package cache records for: {', '.join(missing)}"
        )
    return existing
```

Reply:
cool, I'll revert the commit that pulled in this change